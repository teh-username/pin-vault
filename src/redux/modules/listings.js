import { combineReducers } from 'redux';

export const ADD_ENTRY = 'modules/listings/ADD_ENTRY';
export const MODIFY_ENTRY = 'modules/listings/MODIFY_ENTRY';
export const DELETE_ENTRY = 'modules/listings/DELETE_ENTRY';

/*
  Sample state:
  {
    entries: {
      2cf05d: {
        name: 'Bank 1',
        code: '1234',
      },
      5d94db: {
        name: 'Bank 2',
        code: '5678',
      },
    },
  }
*/

const initialState = {
  entries: {},
};

export const entries = (state = initialState.entries, action) => {
  switch (action.type) {
    case ADD_ENTRY:
      return {
        ...state,
        [action.id]: {
          code: action.code,
          name: action.name,
        },
      };
    case MODIFY_ENTRY:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          code: action.code,
        },
      };
    case DELETE_ENTRY:
      return Object.entries(state)
        .filter(([id, entryDetails]) => action.id !== id)
        .reduce((previousState, [id, entryDetails]) => {
          return {
            ...previousState,
            [id]: entryDetails,
          };
        }, {});
    default:
      return state;
  }
};

export default combineReducers({
  entries,
});

export const addEntry = (id, name, code) => ({
  type: ADD_ENTRY,
  id,
  name,
  code,
});

export const modifyEntry = (id, code) => ({
  type: MODIFY_ENTRY,
  id,
  code,
});

export const deleteEntry = id => ({
  type: DELETE_ENTRY,
  id,
});

export const getEntries = ({ listings: { entries } }) =>
  Object.entries(entries);
export const getEntryDetails = ({ listings: { entries } }, id) => entries[id];
