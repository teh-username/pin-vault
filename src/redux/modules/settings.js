import { combineReducers } from 'redux';

export const TOGGLE_PASSCODE_REQUIREMENT =
  'modules/settings/TOGGLE_PASSCODE_REQUIREMENT';
export const SET_PASSCODE = 'modules/settings/SET_PASSCODE';

/*
  Sample state:
  {
    requirePasscode:  true,
    passcodeHash: '5E884898DA28047151D0E56F8DC6292773603D0D6AABBDD62A11EF721D1542D8',
  }
 */

const initialState = {
  requirePasscode: false,
  passcodeHash: null,
};

const requirePasscode = (state = initialState.requirePasscode, action) => {
  switch (action.type) {
    case TOGGLE_PASSCODE_REQUIREMENT: {
      return !state;
    }
    default:
      return state;
  }
};

const passcodeHash = (state = initialState.passcodeHash, action) => {
  switch (action.type) {
    case SET_PASSCODE:
      return action.passcodeHash;
    default:
      return state;
  }
};

export const togglePasscodeRequirement = () => ({
  type: TOGGLE_PASSCODE_REQUIREMENT,
});

export const setPasscode = passcodeHash => ({
  type: SET_PASSCODE,
  passcodeHash,
});

export default combineReducers({
  requirePasscode,
  passcodeHash,
});

export const getCurrentPasscode = ({ settings: { passcodeHash } }) =>
  passcodeHash;

export const getPasscodeRequirement = ({ settings: { requirePasscode } }) =>
  requirePasscode;
