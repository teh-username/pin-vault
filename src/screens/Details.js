import React from 'react';
import { Alert } from 'react-native';
import { connect } from 'react-redux';
import EntryDetail from '../components/EntryDetail';

import {
  MODIFY_ENTRY,
  getEntryDetails,
  deleteEntry,
} from '../redux/modules/listings';

export class Details extends React.Component {
  static navigationOptions = {
    title: 'Details',
  };

  constructor(props) {
    super(props);
    this.handleDeletionPress = this.handleDeletionPress.bind(
      this,
      props.navigation.state.params.id
    );
    this.handleEditPress = this.handleEditPress.bind(
      this,
      props.navigation.state.params.id
    );
  }

  handleEditPress(id) {
    this.props.navigation.navigate('Entry', { action: MODIFY_ENTRY, id });
  }

  handleDeletionPress(id) {
    Alert.alert(
      'Entry Deletion',
      'Are you sure? This is irreversible.',
      [
        {
          text: 'Cancel',
          style: 'cancel',
        },
        {
          text: 'Delete',
          onPress: () => {
            this.props.handleDeleteEntry(id);
            this.props.navigation.popToTop();
          },
          style: 'destructive',
        },
      ],
      { cancelable: false }
    );
  }

  render() {
    const { entryDetails: { name, code } } = this.props;
    return (
      <EntryDetail
        name={name}
        code={code}
        onEditPress={this.handleEditPress}
        onDeletePress={this.handleDeletionPress}
      />
    );
  }
}

const mapDispatchToProps = dispatch => ({
  handleDeleteEntry(id) {
    dispatch(deleteEntry(id));
  },
});

const mapStateToProps = (
  state,
  { navigation: { state: { params: { id } } } }
) => ({
  entryDetails: getEntryDetails(state, id) || {},
});

export default connect(mapStateToProps, mapDispatchToProps)(Details);
