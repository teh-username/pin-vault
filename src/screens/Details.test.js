import React from 'react';
import { Alert } from 'react-native';

import { Details } from './Details';
import { MODIFY_ENTRY } from '../redux/modules/listings';

describe('Details screen test', () => {
  const navigationMock = {
    state: {
      params: {
        id: 10,
      },
    },
  };

  const entryDetailsMock = {
    name: 'malak',
    code: '1234',
  };

  it('should render as expected', () => {
    const wrapper = shallow(
      <Details navigation={navigationMock} entryDetails={entryDetailsMock} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should navigate to Entry screen and pass the correct params on edit', () => {
    const navMock = {
      ...navigationMock,
      navigate: jest.fn(),
    };
    const wrapper = shallow(
      <Details navigation={navMock} entryDetails={entryDetailsMock} />
    );
    wrapper.instance().handleEditPress();
    expect(navMock.navigate.mock.calls.length).toBe(1);
    expect(navMock.navigate.mock.calls[0][0]).toBe('Entry');
    expect(navMock.navigate.mock.calls[0][1]).toEqual({
      action: MODIFY_ENTRY,
      id: 10,
    });
  });

  it('should delete entry and navigate back to Home on alert click', () => {
    jest.mock('Alert', () => ({
      alert: jest.fn(),
    }));

    const mockDelete = jest.fn();
    const navMock = {
      ...navigationMock,
      popToTop: jest.fn(),
    };

    const wrapper = shallow(
      <Details
        navigation={navMock}
        entryDetails={entryDetailsMock}
        handleDeleteEntry={mockDelete}
      />
    );
    wrapper.instance().handleDeletionPress();
    Alert.alert.mock.calls[0][2][1].onPress();
    expect(mockDelete.mock.calls.length).toBe(1);
    expect(mockDelete.mock.calls[0][0]).toBe(10);
    expect(navMock.popToTop.mock.calls.length).toBe(1);
  });
});
