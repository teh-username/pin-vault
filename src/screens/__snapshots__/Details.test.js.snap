// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details screen test should render as expected 1`] = `
<EntryDetail
  code="1234"
  name="malak"
  onDeletePress={[Function]}
  onEditPress={[Function]}
/>
`;
